akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -775,7 +775,13 @@ private[persistence] trait Eventsourced
             setLastSequenceNr(highestSeqNr)
             _recoveryRunning = false
             try Eventsourced.super.aroundReceive(recoveryBehavior, RecoveryCompleted)
-            finally transitToProcessingState() // in finally in case exception and resume strategy
+            catch {
+              case NonFatal(t) =>
+                try onRecoveryFailure(t, Some(RecoveryCompleted))
+                finally context.stop(self)
+            }
+
+            transitToProcessingState() // in finally in case exception and resume strategy
```

Comment:
This PR makes me realise another thing. We should call `returnRecoveryPermit` inside a `finally` block. We always want to return the permit, not matter what happens.

This `stateReceive` is calling `returnRecoveryPermit` a few times and finishes with 

```scala
catch {
 case NonFatal(e) =>
   returnRecoveryPermit()
   throw e
}
```

Calling it once again if we get a `NonFatal`. 

Instead, we should call it only once and always, doesn't matter if fatal or non-fatal exceptions.

Like in...
```scala
try {
	// do pattern matching and process message
} finally {
   returnRecoveryPermit()
}
```

Anyway, don't need to be included in this PR but feel free to add if you want.

Reply:
Probably not always returnRecoveryPermit() should be called because if I try to wrap everything in finally and call returnRecoveryPermit(), the test "akka.persistence.SnapshotFailureRobustnessSpec.A persistentActor with a failing snapshot must recover state starting from the most recent complete snapshot" is failed. Plus there are case where transitToProcessingState() is not called at all (for example, RecoveryTick). However, I removed transitToProcessingState, mentioned in the first comment.